src/08-lazy-evaluation/lazy-evaluation.md

```diff
@@ -138,123 +135,146 @@ def evalcbn(e: Exp) : Exp = e match {
-*/
+
+```scala mdoc
 val test5res = wth("cons",cons,
                wth("nil", nil,
                  Ap(Ap("cons",2),Ap(Ap("cons",3),Ap(Ap("cons",4), "nil")))))
 assert(eval(test5) == eval(test5res))
-/**
+```
+
 Using evalcbn instead of eval the assertion does not hold (why?), but the results are beta-equivalent.
 We can also construct infinite lists. To this end, we need some form of recursion. We choose the standard fixed-point operator Y.
 This operator only works under call-by-name or other so-called "non-strict" evaluation strategies.
-*/
 
+```scala mdoc
 val y = Fun("f", Ap(Fun("x",Ap("f", Ap("x","x"))), Fun("x",Ap("f",Ap("x","x")))))
+```
 
-/**
 Using Y, we can construct infinite lists, such as the list of all natural numbers.
-*/
+
+```scala mdoc
 val allnats = Ap(Ap("y", Fun("nats", Fun("n", Ap(Ap("cons","n"), Ap("nats", Add("n",1)))))),1)
+```
 
-/**
 We can also perform standard computations on infinite lists, such as mapping the successor function over it.
-*/
 
+```scala mdoc
 val list2toinfty = wth("cons",cons,
                    wth("nil", nil,
                    wth("y", y,
                    wth("maplist", maplist,
                       Ap(Ap("maplist", Fun("x", Add("x",1))), allnats)))))
-/**
+```
 Of course, ``list2toinfty`` diverges when we use ``eval``, but it works fine with ``evalcbn``. It is hard to verify the result due to
 an almost unreadable output. Hence we propose the following
+
 Exercise: Extend the language such that you can implement the "take" function as known from Haskell within the language
 (if0-expressions or something like it are needed for that). Now add a "print" function that prints a number on the console.
 Use it to display the first 3 list elements of test2toinfty are 2,3,4.
--- end of extra material --
-*/
 
-/**
-Environment-based lazy evaluation
----------------------------------
+_end of extra material_
+
+## Environment-based lazy evaluation
+
 Let us now consider the question how we can implement call-by-name in the environment-based interpreter. Translating the idea of not
 evaluating the function argument to the environment-based version seems to  suggest that the environment should map identifiers to
 expression instead of values.
+
 However, we run into the same problems that we had with first-class functions before we introduced closures: What happens to the
 deferred substitutions that still have to be applied in the function argument? If we discard the environment in which the function
 argument was defined we again introduce a variant of dynamic scoping.
+
 Hence, like for closures, we need to store the environment together with the expression. We call such a  pair a _thunk_. An environment
 hence becomes a mapping from symbols to thunks. Note that environments and thunks are hence mutually recursive. In Scala, we can hence
 not use type definitions of the form
+
+```scala
```

Comment:
Here I would leave the Scala out entirely, since this should not end up in the Scala file.